designs/rerun_cli_commands.md

```diff
@@ -0,0 +1,85 @@
+# Rerun CLI commands after failure
+
+## Introduction
+**Problem:**
+Most of the errors returned by the CLI cluster commands are either transient or fixable with manual intervention.
+However, the CLI commands are not idempotent.
+This makes impossible to rerun commands after they have already failed, even if the root cause of the issues has already been solved.
+This makes the experience very disruptive, specially for the upgrade cluster command, which could leave the cluster in an irrecoverable state without the necessary manual cleanup.
+This "cleanup" is not documented and it heavily dependents on the internal CLI's implementation.
+For certain scenarios, it is not even possible, requiring users to destroy and recreate their clusters from scratch.
+
+## Solution
+The proposed solution is to implement a "checkpoint" capability.
```

Comment:
How would a user enable this checkpoint capability? I am assuming that if they want to rerun it based on an unfinished execution they would specify using a flag or something.